### ISSUE TITLE
Using only the backlinks feature in existing RST prose

### ISSUE THREAD

**Issue by felix-hilden** [sphinx-gallery/sphinx-gallery#846]
Hi!

Firstly, thanks for the seemingly great package 😄 I see lots of successful libraries are using it to a great effect. And I intend to be your next user, especially once I write a bunch of python examples. For now though, I'd like to only use the *awesome* backlink feature to enhance my examples that are written inside existing RST prose.

Essentially, what I'm after is being able to write ordinary Sphinx RST like

```rst
Document
========
This is a document with the following example:

.. code:: python

    import lib

    lib.foo()
    b = lib.Bar()
    b.meth()
```

and have the links be generated regardless of the file's location. It's not obvious whether or not this is even possible since the documentation is so focused on the example gallery (which is hardly surprising 😅), but I suspect it is some combination of `backreferences_dir`, `doc_module` and `reference_url` as laid out in the few tutorials starting [here](https://sphinx-gallery.github.io/stable/configuration.html#add-intersphinx-links-to-your-examples). How could this be achieved?

Apologies for essentially asking the library to do something that it wasn't necessarily supposed to do. So feel free to shut this down if you already know it won't be possible. But I'm over the moon that this is even a thing! Keep making documentation great ❤️

---

Eventually having mini-galleries or just plain lists of example locations in reference documentation would be great too, but I think that's too much for one user support question 😅 I might still be misunderstanding the features altogether.

**Comment by lucyleeow** [comment]
This is slightly related to #330 (in particular @GaelVaroquaux's case where we convert .rst to .py so sphinx-gallery can convert back to rst), do you think this would solve your problem? With regards to the .rst file being anywhere - that's possibly a bit tricker, we could enable sphinx-gallery to look through all the `docs/` source dir and convert all rst to .py but I am not sure that is what you want? I am also hesitant about this I don't want to accidently mess up the `docs/` dir structure. @larsoner may have an opinion on this, I don't use Sphinx enough.

(also I see that I said I will take a look at implementing this 😓 - whoops, adding to my todo list)

**Comment by felix-hilden** [comment]
Thanks for the fast reply!

That sounds like it could do the job, but it feels *very* hacky. I partly expected this to not be possible right away. I think don't bother going into implementing just yet 😅 What do you think about extracting the backlink capabilities, either to their own component working independently in this library or to a dedicated Sphinx extension that perhaps you'd use too? If your time is limited, I could have a go at that as well! This feature really seems like it should be in every library's documentation. It only makes things a hundred times better 😄

**Comment by larsoner** [comment]
> What do you think about extracting the backlink capabilities, either to their own component working independently in this library or to a dedicated Sphinx extension that perhaps you'd use too?

The tricky part is that we figure out the backlinking during the execution of the code. So to use our magic, these code blocks would actually need to be executed. This is maybe okay in some circumstances and not others, so 1) it would have to be opt-in in some way and 2) it might end up being a bit of work.

However, it might be worth it. See for example https://github.com/pyvista/pyvista/pull/1457 where PyVista made it so that code in the Examples block gets executed + embedded -- we could have maybe saved them some work (and got backlinking and such for free in the process) if we had implemented this. There is potentially also some overlap with matplotlib's plot directive, and numpydoc's integration with it (which allows for executing code blocks in `examples` and embedding any plot results IIRC).

TL;DR: I agree it's very likely useful, but we should keep in mind the larger context of potential users and use cases, esp. WRT `Examples` sections, and probably talk to folks working with other libraries (numpydoc/matplotlib/pyvista, etc.) about it

**Comment by felix-hilden** [comment]
Totally agree, and thanks for providing the bigger picture!

Running the code is an important point. I think it's definitely possible to statically analyse the code to determine imports and sources. I've done some AST diving myself, mixed with variable and import tracing [here](https://pyfactor.readthedocs.io/en/stable/). I bet it wouldn't be too hard to extend and modify it.

I'm glad you agree that it would be useful 😃 I'm a 100 % sure it would be worth it, but are you indeed proposing for it to be a part of Sphinx Gallery? Honestly, now that I've had some time to think about it I feel like it *shouldn't* be your responsibility, nor should users install a seemingly unrelated package just for the specific functionality. Maybe another, more focused package would be appropriate. Although I'm all for enhancing Sphinx Gallery! What do you think?

Running the code is no doubt important for you though. So are the example sources you generate analysed and executed once they are already RST, or in their `.py` form? If RST, then deciding on the run vs. static analysis (or both as an option toggle) could be easier. It would really be awesome if the functionality I'm after (without galleries) would overlap enough with your goals so that we could work together! Damn I'm still so excited 😄

**Comment by lucyleeow** [comment]
> So are the example sources you generate analysed and executed once they are already RST, or in their .py form?

The example `.py` files are split into text (rst) and code sections, (called 'blocks' in the code), code sections are run, the output is captured and then text sections, along with rst-ed code and ouput are put back to give you the rst, which is used by Sphinx.

> So to use our magic, these code blocks would actually need to be executed.

I'm not as familiar with our `backreferences.py` but could execution and static checking of the code be separate? If it can then we can always do the backreferences statically and optionally execute.

**Comment by felix-hilden** [comment]
Makes sense, thanks for clarifying 👍 I suspect it wouldn't be hard to read and concatenate code examples from RST to produce the running code either.

**Comment by lucyleeow** [comment]
>  I suspect it wouldn't be hard to read and concatenate code examples from RST to produce the running code either.

Sorry can you clarify, I didn't follow, thanks!

**Comment by felix-hilden** [comment]
Sure! Given that you execute the `.py` file and then make the RST file with its comment blocks, and the use case of much ordinary documentation which is RST interleaved with code blocks, I think that there would be common ground if you gathered the code only after transforming it to RST. Then the file conversion would be a separate concern from execution and backlinking, and your users could even write RST examples!

**Comment by lucyleeow** [comment]
Are you saying: `.py` example with rST and code blocks -> all rST (same starting point as a pure .rst file) -> if required, take out the code parts and execute them along with backlinking?

I'm not sure how much more efficient that would be vs:
rST -> .py (same starting point as .py file) -> run as usual.

The thing with the first option is that we would have to update the rST after executing to insert the captured output from executing the code blocks back and ensure we put them in the correct place.

**Comment by felix-hilden** [comment]
That's a very good point, I hadn't considered capturing output. My point was simply that if the starting point for you *could* be RST, then a tool capable of generating links from RST could be used by you as well. An RST -> py conversion surely works too 😅 I'm just trying to parse what a component that was usable for both ordinary RST docs (perhaps without execution) and your current case would do exactly.

An RST file can be statically analysed and the links generated on Sphinx build I presume (I haven't actually made custom extensions). But for you it would be best to generate the RST when executing the code, at which point the links would be already generated. They don't show up in the RST source though (I'm guessing the extra files that are generated are to store them). So maybe the generated files (and their persistence) could be the key to a solution that fits both needs. I'm just throwing around anything that comes to my mind at this point though.

**Comment by lucyleeow** [comment]
I think statically getting the backreferences is the tricky part. The tool should be able to work for code blocks regardless of whether the file it came from was .py or .rst -> i.e. it takes a code block and outputs backlined rst.

Whether we go from .py -> rST -> execute/get backreferences OR .py -> execute/get backreferences -> rST , would then be fairly trivial and a decision based on user need and backwards compatibility etc

You can take a look at `[backreferences.py`](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/backreferences.py) to get an idea of how it is done now.

**Comment by felix-hilden** [comment]
I've recently implemented the functionality in a new Sphinx extension [sphinx-codeautolink](https://sphinx-codeautolink.readthedocs.io). As I said, the approach was different enough to warrant a new project in my opinion. Execution is fine for your gallery and actually should work much better in certain situations to determine references. But I went for static analysis instead for fewer assumptions and more flexibility. But I'd like to say thanks for the initial discussion! Your directions and source were very helpful when piecing together what should happen 👍 I'll leave this open if you'd like to do something with it, or if you'd like to discuss something I'm all for it. But you can consider my use case satisfied.

**Comment by lucyleeow** [comment]
Interesting, thanks @felix-hilden ! Maybe we could make a note of this in our docs @larsoner ?

**Comment by felix-hilden** [comment]
Wow that'd be too kind of you 😅

**Comment by larsoner** [comment]
Feel free @lucyleeow !
